columnflow/production/btag.py

```diff
@@ -0,0 +1,197 @@
+
+    as well as an auxiliary entry in the config to refer to the b-tag correction set.
+
+    .. code-block:: python
+
+        cfg.x.btag_sf_correction_set = "deepJet_shape"
+
+    In addition, JEC uncertainty sources are propagated and weight columns are written if an
+    auxiliary config entry ``btag_sf_jec_sources`` exists.
+
+    Resources:
+
+       - https://twiki.cern.ch/twiki/bin/view/CMS/BTagShapeCalibration?rev=26
+       - https://indico.cern.ch/event/1096988/contributions/4615134/attachments/2346047/4000529/Nov21_btaggingSFjsons.pdf
+    """
+    if self.dataset_inst.is_data:
+        return events
+
+    # get the total number of jets in the chunk
+    n_jets_all = len(flat_np_view(events.Jet.pt, axis=1))
+
+    # get flat inputs, evaluated at jet_mask
+    flavor = flat_np_view(events.Jet.hadronFlavour[jet_mask], axis=1)
+    abs_eta = flat_np_view(abs(events.Jet.eta[jet_mask]), axis=1)
+    pt = flat_np_view(events.Jet.pt[jet_mask], axis=1)
+    b_discr = flat_np_view(events.Jet.btagDeepFlavB[jet_mask], axis=1)
+
+    # helper to create and store the weight
+    def add_weight(syst_name, syst_direction, column_name):
+        # define a mask that selects the correct flavor to assign to, depending on the systematic
+        flavor_mask = Ellipsis
+        if syst_name in ["hf", "lf", "hfstats1", "hfstats2", "lfstats1", "lfstats2"]:
+            flavor_mask = flavor != 4
+        elif syst_name in ["cferr1", "cferr2"]:
+            flavor_mask = flavor == 4
+
+        # get the flat scale factors
+        sf_flat = self.btag_sf_corrector.evaluate(
+            syst_name if syst_name == "central" else f"{syst_direction}_{syst_name}",
+            flavor[flavor_mask],
+            abs_eta[flavor_mask],
+            pt[flavor_mask],
+            b_discr[flavor_mask],
+        )
+
+        # insert them into an array of ones whose length corresponds to the total number of jets
+        sf_flat_all = np.ones(n_jets_all, dtype=np.float32)
+        if jet_mask is Ellipsis:
+            indices = flavor_mask
+        else:
+            indices = flat_np_view(jet_mask)
+            if flavor_mask is not Ellipsis:
+                indices = np.where(indices)[0][flavor_mask]
+        sf_flat_all[indices] = sf_flat
+
+        # enforce the correct shape and create the product via all jets per event
+        sf = layout_ak_array(sf_flat_all, events.Jet.pt)
+        weight = ak.prod(sf, axis=1, mask_identity=False)
+
+        # save the new column
+        return set_ak_column(events, column_name, ak.values_astype(weight, np.float32))
+
+    # when the uncertainty is a known jec shift, obtain the propagated effect and do not produce
+    # additional systematics
+    if self.shift_inst.is_nominal:
+        # nominal weight and those of all method intrinsic uncertainties
+        events = add_weight("central", None, "btag_weight")
+        for syst_name, col_name in self.btag_uncs.items():
```

Comment:
Are the `btag_uncs` always the same or should we allow setting them in some config?

Reply:
For the shape calibrating SFs, the uncertainties have been the same since roughly 5 years, and they should not change in the future, so I thought it's ok for have them in there. Also, there are some hard-coded conditions necessary anyway, e.g.

https://github.com/uhh-cms/columnflow/blob/4dadfe46ebf9786fdc6efcf7370b36bb43c05fe3/columnflow/production/btag.py#L66-L69